### PR TITLE
Add StringSegment.Compare and have StringSegmentComparer implement IComparer<StringSegment>

### PR DESCRIPTION
**Pull request by dlstucki**
Add StringSegmentComparer.Compare(StringSegment, StringSegment) support.

### CODE REVIEW

**Comment by dstuckims** [comment]
Would it be more appropriate to have a static Compare method on StringSegment class itself?

``` CSharp
/// <summary>
/// Compares substrings of two specified <see cref="StringSegment"/> objects using the specified rules,
/// and returns an integer that indicates their relative position in the sort order.
/// </summary>
/// <param name="a">The first StringSegment to compare.</param>
/// <param name="b">The second StringSegment to compare.</param>
/// <param name="comparisonType">One of the enumeration values that specifies the rules for the comparison.</param>
/// <returns>
/// A 32-bit signed integer indicating the lexical relationship between the two comparands.
/// The value is negative if <paramref name="a"/> is less than <paramref name="b"/>, 0 if the two comparands are equal,
/// and positive if <paramref name="a"/> is greater than <paramref name="b"/>.
/// </returns>
public static int Compare(StringSegment a, StringSegment b, StringComparison comparisonType)
{
    int minLength = Math.Min(a.Length, b.Length);
    int diff = string.Compare(a.Buffer, a.Offset, b.Buffer, b.Offset, minLength, comparisonType);
    if (diff == 0)
    {
        diff = a.Length - b.Length;
    }

    return diff;
}
```

**Comment by Tratcher** [comment]
Mimic the String/StringComparer APIs wherever it makes sense. These are supposed to be drop in replacements.

File: src/Microsoft.Extensions.Primitives/StringSegmentComparer.cs [hunk 1]
```diff
@@ -21,7 +21,19 @@ private StringSegmentComparer(StringComparison comparison, StringComparer compar
 
         private StringComparison Comparison { get; }
         private StringComparer Comparer { get; }
+		
+        public int Compare(StringSegment x, StringSegment y)
+        {
+            int minLength = Math.Min(x.Length, y.Length);
```
Comment (dstuckims) [inline src/Microsoft.Extensions.Primitives/StringSegmentComparer.cs]: ok, it's 3 characters either way, had selected the more informative option...

Comment (dstuckims) [inline src/Microsoft.Extensions.Primitives/StringSegmentComparer.cs]: Find updated code in StringSegment.cs in 3rd update.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 2]
```diff
@@ -660,6 +660,49 @@ public void StringSegment_Subsegment_InvalidOffsetAndLength()
             // Act & Assert
             Assert.Throws<ArgumentOutOfRangeException>(() => segment.Subsegment(2, 3));
         }
+		
+        public static TheoryData<StringSegment, StringSegmentComparer, int> CompareStringSegmentData
```
Comment (dstuckims) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Addressed in the 3rd push.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 3]
```diff
@@ -660,6 +660,49 @@ public void StringSegment_Subsegment_InvalidOffsetAndLength()
...
+
+        [Theory]
+        [MemberData(nameof(CompareStringSegmentData))]
+        public void StringSegment_Compare_Valid(StringSegment candidate, StringSegmentComparer comparer, int expectedResult)
+        {
+            // Arrange
+            var segment = new StringSegment("abcdef", 1, 4);
+
+            // Act
+            int result = comparer.Compare(segment, candidate);
```
Comment (dstuckims) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Addressed in the 3rd push.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 4]
```diff
@@ -660,6 +660,49 @@ public void StringSegment_Subsegment_InvalidOffsetAndLength()
...
+
+            // Assert
+            if (expectedResult < 0)
```
Comment (dstuckims) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Addressed in the 3rd push.

Comment (dstuckims) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Addressed in the 3rd push.

File: src/Microsoft.Extensions.Primitives/StringSegmentComparer.cs (hunk 1 above)
Comment (dstuckims) [inline src/Microsoft.Extensions.Primitives/StringSegmentComparer.cs]: Addressed in the 3rd push.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 5]
```diff
@@ -661,6 +661,91 @@ public void StringSegment_Subsegment_InvalidOffsetAndLength()
             Assert.Throws<ArgumentOutOfRangeException>(() => segment.Subsegment(2, 3));
         }
```
Comment (Tratcher) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: could use a test case for null or empty segments.

**Comment by aspnet-hello** [comment]
This issue was moved to aspnet/Home#2364

**Comment by davidfowl** [comment]
@pranavkm @Tratcher any more feedback? 

@dlstucki can you rebase please? (Sorry for the delay)

**Comment by pranavkm** [comment]
Nope, I signed off on this a while ago

**Comment by pranavkm** [comment]
Merged in https://github.com/aspnet/Common/commit/6dd6a2551c2655a7d1024b58a1567aa01f40141a. Thanks for the PR!

**Pull request closed without merging**